File: Java/src/main/java/com/gildedrose/GildedRose.java
package com.gildedrose;

import java.util.Arrays;

class GildedRose {
    Item[] items;

    public GildedRose(Item... items) {
        this.items = items;
    }

    public void updateQuality() {
        this.items = Arrays.stream(items)
                .map(item -> item.name.equals("Sulfuras, Hand of Ragnaros")
                        ? new Item(item.name, item.sellIn, item.quality)
                        : new Item(item.name, item.sellIn - 1, Math.min(50, Math.max(0, item.quality - degradation(item)))))
                .toArray(Item[]::new);
    }

    private int degradation(Item item) {
        if ("Aged Brie".equals(item.name)) return -defaultDegradation(item);
        else if ("Backstage passes to a TAFKAL80ETC concert".equals(item.name)) return backstageDegradation(item);
        else if ("Conjured Mana Cake".equals(item.name)) return 2 * defaultDegradation(item);
        else return defaultDegradation(item);
    }

    private int defaultDegradation(Item item) {
        return item.sellIn > 0 ? 1 : 2;
    }

    private int backstageDegradation(Item item) {
        if (item.sellIn <= 0) return -item.quality;
        else if (item.sellIn <= 5) return -3;
        else if (item.sellIn <= 10) return -2;
        else return -1;
    }
}

File: Kotlin/src/main/kotlin/com/gildedrose/GildedRose.kt
package com.gildedrose

import java.lang.Integer.*

private const val QUALITY_MIH = 0
private const val QUALITY_MAX = 50
private fun Int.degrade(step: Int = 1) = min(max(this - step, QUALITY_MIH), QUALITY_MAX)

class GildedRose(var items: Array<Item>) {

    private val strategies = mapOf(
        "Aged Brie" to Brie::updateQuality,
        "Backstage passes to a TAFKAL80ETC concert" to BackstagePass::updateQuality,
        "Sulfuras, Hand of Ragnaros" to Sulfuras::updateQuality,
        "Conjured Mana Cake" to Conjured::updateQuality,
    )

    fun updateQuality() {
        items = items.map { (strategies[it.name] ?: Normal::updateQuality).invoke(it) }.toTypedArray()
    }

    abstract class AbstractStrategy {
        open fun Item.advance() = sellIn - 1
        open fun updateQuality(item: Item) =
            item.advance()
                .takeIf { sellIn -> sellIn != item.sellIn }
                ?.let { Item(item.name, it, item.quality.degrade(degradation(it))) }
                ?: item

        open fun degradation(sellIn: Int) = 0
    }

    private object Normal : AbstractStrategy() {
        override fun degradation(sellIn: Int) = if (sellIn >= 0) 1 else 2
    }

    private object Brie : AbstractStrategy() {
        override fun degradation(sellIn: Int) = if (sellIn >= 0) -1 else -2
    }

    private object BackstagePass : AbstractStrategy() {
        override fun updateQuality(item: Item) =
            item.advance().let {
                Item(item.name, it, if (it < 0) 0 else item.quality.degrade(degradation(it)))
            }

        override fun degradation(sellIn: Int) = when {
            sellIn < 5 -> -3
            sellIn < 10 -> -2
            else -> -1
        }
    }

    private object Sulfuras : AbstractStrategy() {
        override fun Item.advance() = 0
    }

    private object Conjured : AbstractStrategy() {
        override fun degradation(sellIn: Int) = if (sellIn >= 0) 2 else 4
    }
}
